File: Sqrt.py
"""
Implement int sqrt(int x).

Compute and return the square root of x.
"""


class Solution:
    # @param {integer} x
    # @return {integer}
    def mySqrt(self, x):
        left, right = 0, x
        while left < right:
            mid = left + (right - left) / 2
            if mid ** 2 == x or mid ** 2 < x and (mid + 1) ** 2 > x:
                return mid
            if mid ** 2 > x:
                right = mid - 1
            else:
                left = mid + 1
        return left


if __name__ == '__main__':
    s = Solution()
    print map(s.mySqrt, range(10))

File: Leetcode-Java/Sqrt.java
/**
 * Created by yhf on 15/5/24.
 */

/*
Implement int sqrt(int x).

Compute and return the square root of x.
*/

public class Sqrt {
    public static int mySqrt(int x) {
        // 注意int改为long，不然Time Limit Exceeded
        long left = 1;
        long right = x;
        while (left <= right) {
            long mid = left + (right - left) / 2;
            long square = mid * mid;
            if (square == x) {
                return (int) mid;
            }
            if (square > x) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return (int) right;
    }

    public static void main(String[] args) {
        for (int i = 0; i <= 10; i++) {
            System.out.print(i);
            System.out.println(mySqrt(i));
        }
    }
}
